### PR TITLE
mds: track dirty dentries in separate list

### PR DESCRIPTION
**Pull request by ukernel**
this should improve performance of large directory

Fixes: http://tracker.ceph.com/issues/19578
Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

### CODE REVIEW

File: src/mds/CDir.cc [hunk 1]
```diff
@@ -2342,10 +2354,9 @@ void CDir::_committed(int r, version_t v)
     mark_clean();
 
   // dentries clean?
-  for (map_t::iterator it = items.begin();
-       it != items.end(); ) {
-    CDentry *dn = it->second;
-    ++it;
+  for (auto p = dirty_dentries.begin(); !p.end(); ) {
```
Comment (batrick) [inline src/mds/CDir.cc]: We should be able to remove `is_dirty` checks below, yes?

Comment (ukernel) [reply src/mds/CDir.cc]: yes, but only for dn. inode created by MDCache::cow_inode is not dirty

**Comment by ukernel** [comment]
this one is buggy. it caused
http://qa-proxy.ceph.com/teuthology/zyan-2017-12-20_02:35:39-fs-wip-zyan-integration-20171219-testing-basic-mira/1983268/

File: src/mds/CDir.cc [hunk 2]
```diff
@@ -2174,6 +2170,22 @@ void CDir::_omap_commit(int op_prio)
       to_set.clear();
       to_remove.clear();
     }
+  };
+
+  if (state_test(CDir::STATE_DNPINNEDFRAG)) {
```
Comment (ukernel) [inline src/mds/CDir.cc]: should test "state_test(CDir::STATE_FRAGMENTING)"

**Pull request merged**